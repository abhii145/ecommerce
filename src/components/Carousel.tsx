import { useState, useEffect, useCallback } from "react"
import { motion } from "framer-motion"
import { Link } from "react-router"

interface Slide {
  type: string
  background?: string
  imageSrc?: string
  title?: string
  description?: string
  buttonText?: string
  link?: string
}

interface CarouselProps {
  slides: Slide[]
}

const Carousel = ({ slides }: CarouselProps) => {
  const [currentSlide, setCurrentSlide] = useState(0)

  const handleNext = useCallback(() => {
    setCurrentSlide((prev) => (prev + 1) % slides.length)
  }, [slides.length])

  const handlePrev = useCallback(() => {
    setCurrentSlide((prev) => (prev - 1 + slides.length) % slides.length)
  }, [slides.length])

  useEffect(() => {
    const interval = setInterval(() => {
      handleNext()
    }, 5000)

    return () => clearInterval(interval)
  }, [handleNext])

  return (
    <div className="relative" aria-roledescription="carousel">
      <motion.div
        className={`relative ${
          slides[currentSlide].type === "gradient"
            ? slides[currentSlide].background
            : ""
        } text-white`}
        style={
          slides[currentSlide].type === "image"
            ? {
                backgroundImage: `url(${slides[currentSlide].imageSrc})`,
                backgroundSize: "contain",
                backgroundRepeat: "no-repeat",
                backgroundPosition: "center",
                height: "500px",
              }
            : { height: "500px" }
        }
        initial={{ opacity: 0, scale: 0.9 }}
        animate={{ opacity: 1, scale: 1 }}
        transition={{ duration: 0.8 }}
        role="group"
        aria-roledescription="slide"
        aria-label={`Slide ${currentSlide + 1} of ${slides.length}`}
      >
        <div className="container mx-auto text-center h-full flex flex-col justify-center items-center">
          {slides[currentSlide].type === "gradient" && (
            <>
              <h1 className="text-5xl font-bold">
                {slides[currentSlide].title}
              </h1>
              <p className="mt-4 text-lg max-w-md mx-auto">
                {slides[currentSlide].description}
              </p>
            </>
          )}

          <button className="mt-6 px-6 py-2 bg-white text-blue-600 rounded-lg shadow hover:bg-blue-100">
            <Link to={slides[currentSlide].link}>
              {slides[currentSlide].buttonText}
            </Link>
          </button>
        </div>
      </motion.div>

      <div className="absolute top-1/2 left-0 transform -translate-y-1/2 px-4">
        <button
          className="text-white text-2xl bg-black bg-opacity-50 p-2 rounded-full"
          onClick={handlePrev}
          aria-label="Previous Slide"
        >
          &#8249;
        </button>
      </div>
      <div className="absolute top-1/2 right-0 transform -translate-y-1/2 px-4">
        <button
          className="text-white text-2xl bg-black bg-opacity-50 p-2 rounded-full"
          onClick={handleNext}
          aria-label="Next Slide"
        >
          &#8250;
        </button>
      </div>

      <div className="absolute bottom-4 left-1/2 transform -translate-x-1/2 flex space-x-2">
        {slides.map((_, index) => (
          <button
            key={index}
            className={`w-3 h-3 rounded-full ${
              index === currentSlide ? "bg-green-500" : "bg-gray-600"
            }`}
            onClick={() => setCurrentSlide(index)}
            aria-label={`Go to slide ${index + 1}`}
          />
        ))}
      </div>
    </div>
  )
}

export default Carousel
