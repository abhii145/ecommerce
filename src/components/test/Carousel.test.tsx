import { render, screen, fireEvent } from "@testing-library/react"
import Carousel from "../Carousel"
import { BrowserRouter as Router } from "react-router"

const slides = [
  {
    type: "gradient",
    background: "bg-gradient-to-r from-purple-400 via-pink-500 to-red-500",
    title: "Slide 1",
    description: "Description 1",
    buttonText: "Button 1",
    link: "/link1",
  },
  {
    type: "image",
    imageSrc: "https://via.placeholder.com/800x400",
    title: "Slide 2",
    description: "Description 2",
    buttonText: "Button 2",
    link: "/link2",
  },
]

describe("Carousel", () => {
  it("renders the first slide initially", () => {
    render(
      <Router>
        <Carousel slides={slides} />
      </Router>
    )
    expect(screen.getByText("Slide 1")).toBeInTheDocument()
    expect(screen.getByText("Description 1")).toBeInTheDocument()
  })

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Router>
        <Carousel slides={slides} />
      </Router>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("navigates to the previous slide", () => {
    render(
      <Router>
        <Carousel slides={slides} />
      </Router>
    )
    fireEvent.click(screen.getByLabelText("Next Slide"))
    fireEvent.click(screen.getByLabelText("Previous Slide"))
    expect(screen.getByText("Slide 1")).toBeInTheDocument()
    expect(screen.getByText("Description 1")).toBeInTheDocument()
  })
})
