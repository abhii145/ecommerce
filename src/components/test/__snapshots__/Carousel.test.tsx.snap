// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-roledescription="carousel"
    class="relative"
  >
    <div
      aria-label="Slide 1 of 2"
      aria-roledescription="slide"
      class="relative bg-gradient-to-r from-purple-400 via-pink-500 to-red-500 text-white"
      role="group"
      style="height: 500px; opacity: 0; transform: scale(0.9);"
    >
      <div
        class="container mx-auto text-center h-full flex flex-col justify-center items-center"
      >
        <h1
          class="text-5xl font-bold"
        >
          Slide 1
        </h1>
        <p
          class="mt-4 text-lg max-w-md mx-auto"
        >
          Description 1
        </p>
        <button
          class="mt-6 px-6 py-2 bg-white text-blue-600 rounded-lg shadow hover:bg-blue-100"
        >
          <a
            data-discover="true"
            href="/link1"
          >
            Button 1
          </a>
        </button>
      </div>
    </div>
    <div
      class="absolute top-1/2 left-0 transform -translate-y-1/2 px-4"
    >
      <button
        aria-label="Previous Slide"
        class="text-white text-2xl bg-black bg-opacity-50 p-2 rounded-full"
      >
        ‹
      </button>
    </div>
    <div
      class="absolute top-1/2 right-0 transform -translate-y-1/2 px-4"
    >
      <button
        aria-label="Next Slide"
        class="text-white text-2xl bg-black bg-opacity-50 p-2 rounded-full"
      >
        ›
      </button>
    </div>
    <div
      class="absolute bottom-4 left-1/2 transform -translate-x-1/2 flex space-x-2"
    >
      <button
        aria-label="Go to slide 1"
        class="w-3 h-3 rounded-full bg-green-500"
      />
      <button
        aria-label="Go to slide 2"
        class="w-3 h-3 rounded-full bg-gray-600"
      />
    </div>
  </div>
</DocumentFragment>
`;
